File: radialbuttonmenu/src/main/java/huberd/laurent/radialbuttonmenu/RadialMenu.java
package huberd.laurent.radialbuttonmenu;

import android.graphics.Rect;
import android.os.Handler;
import android.support.design.widget.FloatingActionButton;
import android.view.MotionEvent;
import android.view.View;

import java.util.ArrayList;

import huberd.laurent.radialbuttonmenu.Animation.Animation;
import huberd.laurent.radialbuttonmenu.Animation.AnimationManager;
import huberd.laurent.radialbuttonmenu.Translation.LayoutManager;
import huberd.laurent.radialbuttonmenu.Translation.Translation;

public class RadialMenu {

    // region Member variables and Accessors

    private FloatingActionButton mainButton;
    protected ArrayList<FloatingActionButton> secondaryButtons = new ArrayList<>();

    protected RadialMenuState radialMenuState = RadialMenuState.MINIMIZED;

    private TriggerMode triggerMode = TriggerMode.Touch;
    protected LayoutManager layoutManager;
    protected AnimationManager animationManager;

    // Used to enlarge the Radial Menu's bounds in isTouchInBounds method
    protected float topPadding;
    protected float rightPadding;
    protected float bottomPadding;
    protected float leftPadding;

    protected OnRadialMenuStateChangedListener onRadialMenuStateChangeListener;
    protected OnRadialMenuInteractionListener onRadialMenuInteractionListener;

    // endregion

    // region Constructor

    public RadialMenu(FloatingActionButton mainFloatingActionButton, LayoutManager layoutManager, AnimationManager animationManager) {
        this.mainButton = mainFloatingActionButton;
        this.layoutManager = layoutManager;
        this.animationManager = animationManager;

        this.mainButton.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent motionEvent) {
                return handleMainButtonTouch(motionEvent);
            }
        });
    }

    // endregion

    // region Public methods

    public void addSecondaryButton(FloatingActionButton floatingActionButton) {
        configureNewButton(floatingActionButton);
        secondaryButtons.add(floatingActionButton);
    }

    public void addSecondaryButton(int index, FloatingActionButton floatingActionButton) {
        configureNewButton(floatingActionButton);
        secondaryButtons.add(index, floatingActionButton);
    }

    public void removeSecondaryButton(FloatingActionButton floatingActionButton) {
        secondaryButtons.remove(floatingActionButton);
        returnToDefaultPosition(floatingActionButton);
    }

    public int secondaryButtonsSize() {
        return secondaryButtons.size();
    }

    public void setTriggerMode(TriggerMode triggerMode) {
        this.triggerMode = triggerMode;
    }

    public RadialMenuState getRadialMenuState() {
        return radialMenuState;
    }

    public void setOnRadialMenuStateChangeListener(OnRadialMenuStateChangedListener listener) {
        this.onRadialMenuStateChangeListener = listener;
    }

    public void setOnRadialMenuInteractionListener(OnRadialMenuInteractionListener listener) {
        this.onRadialMenuInteractionListener = listener;
    }

    public void hide() {
        if (radialMenuState != RadialMenuState.EXTENDED) return;

        RadialMenuState oldState = radialMenuState;
        radialMenuState = RadialMenuState.ANIMATING;

        if (onRadialMenuStateChangeListener != null) {
            onRadialMenuStateChangeListener.onStateChanged(oldState, RadialMenuState.ANIMATING);
        }

        Translation translation = layoutManager.getMainButtonTranslation(this, mainButton);
        Animation animation = animationManager.getHideMainButtonAnimation(mainButton, translation);

        applyTransformationAndAnimation(mainButton, translation.reverse(), animation, new Runnable() {
            @Override
            public void run() {
                returnToDefaultPosition(mainButton);

                RadialMenuState oldState = radialMenuState;
                radialMenuState = RadialMenuState.MINIMIZED;

                if (onRadialMenuStateChangeListener != null) {
                    onRadialMenuStateChangeListener.onStateChanged(oldState, RadialMenuState.MINIMIZED);
                }

            }
        });

        for (int i = 0; i < secondaryButtons.size(); i++) {
            final FloatingActionButton fab = secondaryButtons.get(i);
            translation = layoutManager.getSecondaryButtonTranslation(this, i, fab);
            animation = animationManager.getHideSecondaryButtonAnimation(i, fab, translation);

            applyTransformationAndAnimation(fab, translation.reverse(), animation, new Runnable() {
                @Override
                public void run() {
                    returnToDefaultPosition(fab);
                }
            });
        }

        resetRadialButtonBounds();
    }

    public void show() {
        if (radialMenuState != RadialMenuState.MINIMIZED) return;

        RadialMenuState oldState = radialMenuState;
        radialMenuState = RadialMenuState.ANIMATING;

        if (onRadialMenuStateChangeListener != null) {
            onRadialMenuStateChangeListener.onStateChanged(oldState, RadialMenuState.ANIMATING);
        }

        Translation translation = layoutManager.getMainButtonTranslation(this, mainButton);
        Animation animation = animationManager.getShowMainButtonAnimation(mainButton, translation);

        applyTransformationAndAnimation(mainButton, translation, animation, null);

        // Used to activate secondary button's interaction before animation finishes.
        // The result is a much easier to use Radial Menu
        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                RadialMenuState oldState = radialMenuState;
                radialMenuState = RadialMenuState.EXTENDED;

                if (onRadialMenuStateChangeListener != null) {
                    onRadialMenuStateChangeListener.onStateChanged(oldState, RadialMenuState.EXTENDED);
                }
            }
        }, animation.getDuration() / 2);

        for (int i = 0; i < secondaryButtons.size(); i++) {
            FloatingActionButton fab = secondaryButtons.get(i);
            translation = layoutManager.getSecondaryButtonTranslation(this, i, fab);
            animation = animationManager.getShowSecondaryButtonAnimation(i, fab, translation);

            applyTransformationAndAnimation(fab, translation, animation, null);
            fab.setVisibility(View.VISIBLE);
            enlargeRadialButtonBounds(translation);
        }
    }

    // endregion

    // region MotionEvent handling

    private boolean handleMainButtonTouch(MotionEvent motionEvent) {
        switch (motionEvent.getAction()) {
            case MotionEvent.ACTION_DOWN:
                handleEventTouchDown(motionEvent);
            case MotionEvent.ACTION_UP:
                handleEventTouchUp(motionEvent);
            case MotionEvent.ACTION_MOVE:
                handleEventTouchMove(motionEvent);
            default:
                return true;
        }
    }

    private boolean handleEventTouchDown(MotionEvent motionEvent) {
        switch (radialMenuState) {
            case ANIMATING:
                return false;
            case MINIMIZED:
                show();
                break;
            case EXTENDED:
                hide();
                break;
        }
        return true;
    }

    private boolean handleEventTouchUp(MotionEvent motionEvent) {
        if (triggerMode == TriggerMode.Touch && radialMenuState == RadialMenuState.EXTENDED) {
            if (isTouchInBounds(motionEvent)) {
                int index = isTouchInSecondaryButtonBounds(motionEvent);
                if (index != -1) {
                    onSecondaryButtonSelected(index);
                    return true;
                }
            } else {
                hide();
            }
        }
        return false;
    }

    private boolean handleEventTouchMove(MotionEvent motionEvent) {
        if (triggerMode == TriggerMode.Touch && radialMenuState == RadialMenuState.EXTENDED) {
            int index = isTouchInSecondaryButtonBounds(motionEvent);
            if (index != -1) {
                onSecondaryButtonHover(index);
                return true;
            }
        }
        return false;
    }

    private void onSecondaryButtonSelected(int i) {
        if (onRadialMenuInteractionListener != null) {
            onRadialMenuInteractionListener.onOptionSelected(i);
        }
        hide();
    }

    private void onSecondaryButtonHover(int i) {
        if (onRadialMenuInteractionListener != null) {
            onRadialMenuInteractionListener.onOptionHover(i);
        }
    }

    //endregion

    // region Helper methods

    private void configureNewButton(final FloatingActionButton floatingActionButton) {
        removeSecondaryButton(floatingActionButton);
        returnToDefaultPosition(floatingActionButton);
        floatingActionButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                int index = secondaryButtons.indexOf(floatingActionButton);
                if (index != -1) {
                    onSecondaryButtonSelected(index);
                }
            }
        });
    }

    private void returnToDefaultPosition(FloatingActionButton floatingActionButton) {
        floatingActionButton.clearAnimation();
        floatingActionButton.setY(mainButton.getY());
        floatingActionButton.setX(mainButton.getX());
        floatingActionButton.setScaleX(1);
        floatingActionButton.setScaleY(1);
        floatingActionButton.setRotation(0);

        mainButton.bringToFront();
    }

    private void applyTransformationAndAnimation(FloatingActionButton floatingActionButton, Translation translation, Animation animation, Runnable endAction) {
        floatingActionButton.clearAnimation();

        floatingActionButton.animate().translationXBy(translation.getX()).translationYBy(translation.getY()).setInterpolator(animation.getTranslationInterpolator()).setDuration(animation.getDuration()).withEndAction(endAction).start();
        floatingActionButton.animate().rotationBy(animation.getRotation()).setInterpolator(animation.getRotationInterpolator()).setDuration(animation.getDuration()).start();
        floatingActionButton.animate().scaleX(animation.getScale()).scaleY(animation.getScale()).setInterpolator(animation.getScaleInterpolator()).setDuration(animation.getDuration()).start();
    }

    private void enlargeRadialButtonBounds(Translation translation) {
        if (translation.getX() > 0) {
            rightPadding = Math.max(rightPadding, translation.getX());
        } else {
            leftPadding = Math.max(leftPadding, Math.abs(translation.getX()));
        }
        if (translation.getY() > 0) {
            bottomPadding = Math.max(bottomPadding, translation.getY());
        } else {
            topPadding = Math.max(topPadding, Math.abs(translation.getY()));
        }
    }

    private void resetRadialButtonBounds() {
        topPadding = 0;
        rightPadding = 0;
        bottomPadding = 0;
        leftPadding = 0;
    }

    private int isTouchInSecondaryButtonBounds(MotionEvent motionEvent) {
        for (int i = 0; i < secondaryButtons.size(); i++) {
            Rect buttonRect = new Rect();
            secondaryButtons.get(i).getGlobalVisibleRect(buttonRect);
            if (buttonRect.contains((int) motionEvent.getRawX(), (int) motionEvent.getRawY())) {
                return i;
            }
        }
        return -1;
    }

    private boolean isTouchInBounds(MotionEvent motionEvent) {
        Rect r = new Rect();
        mainButton.getGlobalVisibleRect(r);
        r.top -= (topPadding + 100);
        r.right += (rightPadding + 100);
        r.bottom += (bottomPadding + 100);
        r.left -= (leftPadding + 100);

        return r.contains((int) motionEvent.getRawX(), (int) motionEvent.getRawY());
    }

    // endregion

    // region RadialMenuState Enum

    public enum RadialMenuState {
        EXTENDED, MINIMIZED, ANIMATING
    }

    // endregion

    // region TriggerMode Enum

    public enum TriggerMode {
        Click, Touch
    }

    // endregion

    // region Interfaces

    public interface OnRadialMenuInteractionListener {
        void onOptionSelected(int i);
        void onOptionHover(int i);
    }

    public interface OnRadialMenuStateChangedListener {
        void onStateChanged(RadialMenuState oldState, RadialMenuState newState);
    }

    // endregion
}

File: radialbuttonmenu/src/main/java/huberd/laurent/radialbuttonmenu/RadialMenu.kt
package huberd.laurent.radialbuttonmenu

import android.graphics.Rect
import android.os.Handler
import android.support.design.widget.FloatingActionButton
import android.view.MotionEvent
import android.view.View

import huberd.laurent.radialbuttonmenu.Animation.Animation
import huberd.laurent.radialbuttonmenu.Animation.AnimationManager
import huberd.laurent.radialbuttonmenu.Translation.LayoutManager
import huberd.laurent.radialbuttonmenu.Translation.Translation
import kotlin.collections.ArrayList

class RadialMenu(
        val mainButton: FloatingActionButton,
        private val secondaryButtons: ArrayList<FloatingActionButton> = ArrayList<FloatingActionButton>(),
        var layoutManager: LayoutManager,
        var animationManager: AnimationManager,
        var onRadialMenuStateChangeListener: OnRadialMenuStateChangedListener? = null,
        var onRadialMenuInteractionListener: OnRadialMenuInteractionListener? = null
) {

    var radialMenuState = RadialMenuState.MINIMIZED
        private set

    var triggerMode = TriggerMode.Touch

    // Used to enlarge the Radial Menu's bounds in isTouchInBounds method
    private var topPadding: Float = 0f
    private var rightPadding: Float = 0f
    private var bottomPadding: Float = 0f
    private var leftPadding: Float = 0f

    init {
        mainButton.setOnTouchListener { _, motionEvent -> handleMainButtonTouch(motionEvent) }
    }

    // region Public methods

    fun addSecondaryButton(floatingActionButton: FloatingActionButton) {
        configureNewButton(floatingActionButton)
        secondaryButtons.add(floatingActionButton)
    }

    fun addSecondaryButton(index: Int, floatingActionButton: FloatingActionButton) {
        configureNewButton(floatingActionButton)
        secondaryButtons.add(index, floatingActionButton)
    }

    fun removeSecondaryButton(floatingActionButton: FloatingActionButton) {
        secondaryButtons.remove(floatingActionButton)
        returnToDefaultPosition(floatingActionButton)
    }

    fun removeSecondaryButton(index: Int, floatingActionButton: FloatingActionButton) {
        secondaryButtons.removeAt(index)
        returnToDefaultPosition(floatingActionButton)
    }

    fun secondaryButtonsSize(): Int {
        return secondaryButtons.size
    }

    fun hide() {
        if (radialMenuState != RadialMenuState.EXTENDED) return

        changeRadialMenuState(RadialMenuState.ANIMATING)

        var translation = layoutManager.getMainButtonTranslation(this, mainButton)
        var animation = animationManager.getHideMainButtonAnimation(mainButton, translation)

        applyTransformationAndAnimation(mainButton, translation.reverse(), animation, Runnable {
            returnToDefaultPosition(mainButton)
            changeRadialMenuState(RadialMenuState.MINIMIZED)
        })

        for (i in secondaryButtons.indices) {
            val fab = secondaryButtons[i]
            translation = layoutManager.getSecondaryButtonTranslation(this, i, fab)
            animation = animationManager.getHideSecondaryButtonAnimation(i, fab, translation)

            applyTransformationAndAnimation(fab, translation.reverse(), animation, Runnable { returnToDefaultPosition(fab) })
        }

        resetRadialButtonBounds()
    }

    fun show() {
        if (radialMenuState != RadialMenuState.MINIMIZED) return

        changeRadialMenuState(RadialMenuState.ANIMATING)

        var translation = layoutManager.getMainButtonTranslation(this, mainButton)
        var animation = animationManager.getShowMainButtonAnimation(mainButton, translation)

        applyTransformationAndAnimation(mainButton, translation, animation, null)

        // Used to activate secondary button's interaction before animation finishes.
        // The result is a much easier to use Radial Menu
        Handler().postDelayed({
            changeRadialMenuState(RadialMenuState.EXTENDED)
        }, (animation.duration / 2).toLong())

        for (i in secondaryButtons.indices) {
            val fab = secondaryButtons[i]
            translation = layoutManager.getSecondaryButtonTranslation(this, i, fab)
            animation = animationManager.getShowSecondaryButtonAnimation(i, fab, translation)

            applyTransformationAndAnimation(fab, translation, animation, null)
            fab.visibility = View.VISIBLE
            enlargeRadialButtonBounds(translation)
        }
    }

    // endregion

    // region MotionEvent handling

    private fun handleMainButtonTouch(motionEvent: MotionEvent): Boolean {
        when (motionEvent.action) {
            MotionEvent.ACTION_DOWN -> {
                handleEventTouchDown(motionEvent)
                handleEventTouchUp(motionEvent)
                handleEventTouchMove(motionEvent)
                return true
            }
            MotionEvent.ACTION_UP -> {
                handleEventTouchUp(motionEvent)
                handleEventTouchMove(motionEvent)
                return true
            }
            MotionEvent.ACTION_MOVE -> {
                handleEventTouchMove(motionEvent)
                return true
            }
            else -> return true
        }
    }

    private fun handleEventTouchDown(motionEvent: MotionEvent): Boolean {
        when (radialMenuState) {
            RadialMenu.RadialMenuState.ANIMATING -> return false
            RadialMenu.RadialMenuState.MINIMIZED -> show()
            RadialMenu.RadialMenuState.EXTENDED -> hide()
        }
        return true
    }

    private fun handleEventTouchUp(motionEvent: MotionEvent): Boolean {
        if (triggerMode == TriggerMode.Touch && radialMenuState == RadialMenuState.EXTENDED) {
            if (isTouchInBounds(motionEvent)) {
                val index = isTouchInSecondaryButtonBounds(motionEvent)
                if (index != -1) {
                    onSecondaryButtonSelected(index)
                    return true
                }
            } else {
                hide()
            }
        }
        return false
    }

    private fun handleEventTouchMove(motionEvent: MotionEvent): Boolean {
        if (triggerMode == TriggerMode.Touch && radialMenuState == RadialMenuState.EXTENDED) {
            val index = isTouchInSecondaryButtonBounds(motionEvent)
            if (index != -1) {
                onSecondaryButtonHover(index)
                return true
            }
        }
        return false
    }

    private fun onSecondaryButtonSelected(i: Int) {
        onRadialMenuInteractionListener?.onOptionSelected(i)
        hide()
    }

    private fun onSecondaryButtonHover(i: Int) {
        onRadialMenuInteractionListener?.onOptionHover(i)
    }

    //endregion

    // region Helper methods

    private fun changeRadialMenuState(state: RadialMenuState) {
        val oldState = radialMenuState
        radialMenuState = state

        onRadialMenuStateChangeListener?.onStateChanged(oldState, RadialMenuState.ANIMATING)
    }

    private fun configureNewButton(floatingActionButton: FloatingActionButton) {
        removeSecondaryButton(floatingActionButton)
        returnToDefaultPosition(floatingActionButton)
        floatingActionButton.setOnClickListener {
            val index = secondaryButtons.indexOf(floatingActionButton)
            if (index != -1) {
                onSecondaryButtonSelected(index)
            }
        }
    }

    private fun returnToDefaultPosition(floatingActionButton: FloatingActionButton) {
        floatingActionButton.clearAnimation()
        floatingActionButton.y = mainButton.y
        floatingActionButton.x = mainButton.x
        floatingActionButton.scaleX = 1f
        floatingActionButton.scaleY = 1f
        floatingActionButton.rotation = 0f

        mainButton.bringToFront()
    }

    private fun applyTransformationAndAnimation(floatingActionButton: FloatingActionButton, translation: Translation, animation: Animation, endAction: Runnable?) {
        floatingActionButton.clearAnimation()

        floatingActionButton.animate().translationXBy(translation.x).translationYBy(translation.y).setInterpolator(animation.translationInterpolator).setDuration(animation.duration.toLong()).withEndAction(endAction).start()
        floatingActionButton.animate().rotationBy(animation.rotation.toFloat()).setInterpolator(animation.rotationInterpolator).setDuration(animation.duration.toLong()).start()
        floatingActionButton.animate().scaleX(animation.scale).scaleY(animation.scale).setInterpolator(animation.scaleInterpolator).setDuration(animation.duration.toLong()).start()
    }

    private fun enlargeRadialButtonBounds(translation: Translation) {
        if (translation.x > 0) {
            rightPadding = Math.max(rightPadding, translation.x)
        } else {
            leftPadding = Math.max(leftPadding, Math.abs(translation.x))
        }
        if (translation.y > 0) {
            bottomPadding = Math.max(bottomPadding, translation.y)
        } else {
            topPadding = Math.max(topPadding, Math.abs(translation.y))
        }
    }

    private fun resetRadialButtonBounds() {
        topPadding = 0f
        rightPadding = 0f
        bottomPadding = 0f
        leftPadding = 0f
    }

    private fun isTouchInSecondaryButtonBounds(motionEvent: MotionEvent): Int {
        for (i in secondaryButtons.indices) {
            val buttonRect = Rect()
            secondaryButtons[i].getGlobalVisibleRect(buttonRect)
            if (buttonRect.contains(motionEvent.rawX.toInt(), motionEvent.rawY.toInt())) {
                return i
            }
        }
        return -1
    }

    private fun isTouchInBounds(motionEvent: MotionEvent): Boolean {
        val r = Rect()
        mainButton.getGlobalVisibleRect(r)
        r.top -= (topPadding + 100).toInt()
        r.right += (rightPadding + 100).toInt()
        r.bottom += (bottomPadding + 100).toInt()
        r.left -= (leftPadding + 100).toInt()

        return r.contains(motionEvent.rawX.toInt(), motionEvent.rawY.toInt())
    }

    // endregion

    // region RadialMenuState Enum

    enum class RadialMenuState {
        EXTENDED, MINIMIZED, ANIMATING
    }

    // endregion

    // region TriggerMode Enum

    enum class TriggerMode {
        Click, Touch
    }

    // endregion

    // region Interfaces

    interface OnRadialMenuInteractionListener {
        fun onOptionSelected(i: Int)
        fun onOptionHover(i: Int)
    }

    interface OnRadialMenuStateChangedListener {
        fun onStateChanged(oldState: RadialMenuState, newState: RadialMenuState)
    }

    // endregion
}
